terraform/transform_output.go

```diff
@@ -37,65 +43,67 @@ func (t *OutputTransformer) transform(g *Graph, c *configs.Config) error {
 		}
 	}
 
-	// Add plannable outputs to the graph, which will be dynamically expanded
+	// Add outputs to the graph, which will be dynamically expanded
 	// into NodeApplyableOutputs to reflect possible expansion
 	// through the presence of "count" or "for_each" on the modules.
-	for _, o := range c.Module.Outputs {
-		node := &nodeExpandOutput{
-			Addr:   addrs.OutputValue{Name: o.Name},
-			Module: c.Path,
-			Config: o,
-		}
-		log.Printf("[TRACE] OutputTransformer: adding %s as %T", o.Name, node)
-		g.Add(node)
-	}
-
-	return nil
-}
-
-// destroyRootOutputTransformer is a GraphTransformer that adds nodes to delete
-// outputs during destroy. We need to do this to ensure that no stale outputs
-// are ever left in the state.
-type destroyRootOutputTransformer struct {
-	Destroy bool
-}
 
-func (t *destroyRootOutputTransformer) Transform(g *Graph) error {
-	// Only clean root outputs on a full destroy
-	if !t.Destroy {
-		return nil
+	// if this is a root output, we add the apply or destroy node directly, as
+	// the root modules does not expand
+	var changes []*plans.OutputChangeSrc
+	if t.Changes != nil {
+		changes = t.Changes.Outputs
 	}
 
-	for _, v := range g.Vertices() {
-		output, ok := v.(*nodeExpandOutput)
-		if !ok {
-			continue
-		}
+	for _, o := range c.Module.Outputs {
+		addr := addrs.OutputValue{Name: o.Name}
 
-		// We only destroy root outputs
-		if !output.Module.Equal(addrs.RootModule) {
-			continue
+		var rootChange *plans.OutputChangeSrc
+		for _, c := range changes {
+			if c.Addr.Module.IsRoot() && c.Addr.OutputValue.Name == o.Name {
+				rootChange = c
+			}
 		}
 
-		// create the destroy node for this output
-		node := &NodeDestroyableOutput{
-			Addr:   output.Addr.Absolute(addrs.RootModuleInstance),
-			Config: output.Config,
+		var node dag.Vertex
+		switch {
+		case c.Path.IsRoot() && t.Destroy:
+			node = &NodeDestroyableOutput{
+				Addr:   addr.Absolute(addrs.RootModuleInstance),
+				Config: o,
+			}
+		case c.Path.IsRoot():
+			destroy := t.Destroy
```

Comment:
we already know destroy is `false` here from the case immediately above, correct? The switch on `case destroy` a couple of lines down seems redundant.

Reply:
Ah yes, I definitely didn't reduce this all the way. Let me pull that `destroy` logic out and remove the nested switch entirely.